File: codesprint/picking.py
from math import factorial;
n=input();

for i in range(0,n):
    k=input();
    a = [0] * k;
    j=0;
    for str in raw_input().split():
        a[j]=long(str);
        j=j+1;
           
    a.sort();
    j=0;
    #for j in range(0,k-1):
    while j<k :
          #print long(a[j]);
          if j<long(a[j]):
              print '0';
              break;
          j=j+1;
    
    if j!=k:
       continue;
          
    f=1;
    t=k-1;
    while t>=0 :
        if t!=k-1 and long(a[t])==long(a[t+1]):
          t=t-1;
          continue;
        temp=t-long(a[t])+1;
        f=f%1000000007*factorial(temp%1000000007)%1000000007;
        t=t-1;
    print f;

File: codesprint/picking.c
#include<stdio.h>
#include<math.h>
long fact(long n)
{
    long k;
    if(n==0)
    return(1);
    else
    {
        k=(n*fact(n-1))%1000000007;
    }
    return(k);
}
int main()
{
long n=0,array[50000],i=0,temp=0,j=0,test=0,k=0,f=1;
scanf("%ld",&test);
for(k=0;k<test;k++)
{
scanf("%ld",&n);

for(i=0;i<n;i++)
{
    scanf("%ld",&array[i]);
}
    for(i=0;i<n;i++)
    {
        for(j=0;j<n-1;j++)
        {
            if(array[i]<array[j])
            {
                temp=array[i];
                array[i]=array[j];
                array[j]=temp;
              }
          }
    }

        for(i=0;i<n;i++)
        {
            if(i<array[i])
            {
                printf("0");
                break;
            }
                        
        }
if(i!=n)
continue;

f=1;
for(i=n-1;i>=0;i--)
{

    if(i!=n-1 && array[i]==array[i+1])
        continue;
    
    temp=i-array[i]+1;
    f=f%1000000007*fact(temp%1000000007)%1000000007;
}
printf("%ld\n",f%1000000007);
}
return 0;
}               

